### PR TITLE
PLT-83: Move the typed validators to plutus-script-utils

### PR DESCRIPTION
**Pull request by andreabedini**
- Move `Ledger.Typed.Scripts.Validators` (plutus-ledger) into `Plutus.Script.Utils.V1.Typed.Scripts.Validators` (plutus-script-utils)
- Move `Ledger.Typed.Tx` (plutus-ledger) into `Plutus.Script.Utils.V1.Typed.Scripts` (plutus-script-utils)
- Move `Ledger.Typed.TypeUtils` (plutus-ledger) into `Plutus.Script.Utils.V1.Typed.TypeUtils` (plutus-script-utils)
- Hide the constructors `TypedScriptTx{In,Out,OutRef}`
- Run pab-nami-demo-generate-purs
- Run plutus-playground-generate-purs
- Remove `ocsTxOut` from `OnChainState`
- Update plutus-contract, plutus-ledger-constraints and plutus-use-cases 
 

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

### CODE REVIEW

File: plutus-ledger/src/Ledger/Typed/Scripts/Validators.hs [hunk 1]
```diff
@@ -1,153 +1,7 @@
...
-        }
-
--- | The minting policy that forwards all checks to the instance's
---   validator
-forwardingMintingPolicy :: TypedValidator a -> Scripts.MintingPolicy
-forwardingMintingPolicy = tvForwardingMPS
+  {-# DEPRECATED "Use Plutus.Script.Utils.V1.Typed.Scripts.Validators instead" #-}
+  ( module Plutus.Script.Utils.V1.Typed.Scripts.Validators,
+  )
+where
```
Comment (koslambrou) [inline plutus-ledger/src/Ledger/Typed/Scripts/Validators.hs:5]: Nice reflex of using deprecation :)

File: plutus-contract/src/Plutus/Contract/StateMachine.hs [hunk 2]
```diff
@@ -422,28 +430,29 @@ runInitialiseWith ::
...
+              <> foldMap (mintingPolicy . curPolicy . ttOutRef) (smThreadToken stateMachine)
+              <> Constraints.unspentOutputs utxo
+              <> customLookups
+      utx <- mapError (review _ConstraintResolutionContractError) (mkTxContract lookups constraints)
+      adjustedUtx <- adjustUnbalancedTx utx
+      unless (utx == adjustedUtx) $
+        logWarn @Text $ "Plutus.Contract.StateMachine.runInitialise: "
+                      <> "Found a transaction output value with less than the minimum amount of Ada. Adjusting ..."
+      submitTxConfirmed adjustedUtx
+      pure initialState
```
Comment (koslambrou) [inline plutus-contract/src/Plutus/Contract/StateMachine.hs:455]: There is no implementation change here right? Just formatting?

File: plutus-contract/src/Plutus/Contract/StateMachine.hs [hunk 3]
```diff
@@ -291,7 +297,9 @@ waitForUpdate ::
     )
     => StateMachineClient state i
     -> Contract w schema e (Maybe (OnChainState state i))
-waitForUpdate client = waitForUpdateTimeout client never >>= awaitPromise >>= \case
+waitForUpdate client = do
+  result <- waitForUpdateTimeout client never >>= awaitPromise
+  case result of
```
Comment (koslambrou) [inline plutus-contract/src/Plutus/Contract/StateMachine.hs:302]: Unnecessary change?

File: plutus-contract/src/Plutus/Contract/StateMachine.hs [hunk 4]
```diff
@@ -261,7 +263,9 @@ waitForUpdateUntilSlot ::
     => StateMachineClient state i
     -> Slot
     -> Contract w schema e (WaitingResult Slot i state)
-waitForUpdateUntilSlot client timeoutSlot = waitForUpdateTimeout client (isSlot timeoutSlot) >>= fmap (fmap (tyTxOutData . ocsTxOut)) . awaitPromise
+waitForUpdateUntilSlot client timeoutSlot = do
+  result <- waitForUpdateTimeout client (isSlot timeoutSlot) >>= awaitPromise
+  pure $ fmap getStateData result
```
Comment (koslambrou) [inline plutus-contract/src/Plutus/Contract/StateMachine.hs:268]: Unnecessary change?

**Review by koslambrou — APPROVED** [review]

File: plutus-contract/src/Plutus/Contract/StateMachine.hs (hunk 2 above)
Comment (andreabedini) [reply plutus-contract/src/Plutus/Contract/StateMachine.hs:455]: Yes, it should have been a separate commit mentioning formatting. The previous line was ticking at 149 characters which I thought was too much to leave it like that.

```
runInitialiseWith customLookups customConstraints StateMachineClient{scInstance} initialState initialValue = mapError (review _SMContractError) $ do
```

Inserting a newline after the `=` indented the rest.

File: plutus-contract/src/Plutus/Contract/StateMachine.hs [hunk 5]
```diff
@@ -274,7 +278,9 @@ waitForUpdateUntilTime ::
     => StateMachineClient state i
     -> POSIXTime
     -> Contract w schema e (WaitingResult POSIXTime i state)
-waitForUpdateUntilTime client timeoutTime = waitForUpdateTimeout client (isTime timeoutTime) >>= fmap (fmap (tyTxOutData . ocsTxOut)) . awaitPromise
+waitForUpdateUntilTime client timeoutTime = do
+  result <- waitForUpdateTimeout client (isTime timeoutTime) >>= awaitPromise
+  pure $ fmap getStateData result
```
[the remark this answers is not in the document]
Comment (andreabedini) [reply plutus-contract/src/Plutus/Contract/StateMachine.hs:283]: Indeed.

The functions`waitForUpdateUntilSlot` and `waitForUpdateUntilTime` were written as super long one-liners. I think splitting them over few lines increses readability. Notice also `tyTxOutData . ocsTxOut` is replaced by `getStateData`.

Before:
```
waitForUpdateUntilSlot client timeoutSlot = waitForUpdateTimeout client (isSlot timeoutSlot) >>= fmap (fmap (tyTxOutData . ocsTxOut)) . awaitPromise

waitForUpdateUntilTime client timeoutTime = waitForUpdateTimeout client (isTime timeoutTime) >>= fmap (fmap (tyTxOutData . ocsTxOut)) . awaitPromise
```

After:
```
waitForUpdateUntilSlot client timeoutSlot = do
  result <- waitForUpdateTimeout client (isSlot timeoutSlot) >>= awaitPromise
  pure $ fmap getStateData result

waitForUpdateUntilTime client timeoutTime = do
  result <- waitForUpdateTimeout client (isTime timeoutTime) >>= awaitPromise
  pure $ fmap getStateData result
```

File: plutus-contract/src/Plutus/Contract/StateMachine.hs (hunk 4 above)
Comment (andreabedini) [reply plutus-contract/src/Plutus/Contract/StateMachine.hs:268]: Answered above.

File: plutus-ledger/src/Ledger/Typed/Scripts/Validators.hs (hunk 1 above)
Comment (andreabedini) [reply plutus-ledger/src/Ledger/Typed/Scripts/Validators.hs:5]: Which triggered warnings in our other packages, which made CI fail :D

File: plutus-contract/src/Plutus/Contract/StateMachine.hs (hunk 3 above)
Comment (andreabedini) [reply plutus-contract/src/Plutus/Contract/StateMachine.hs:302]: This was for consistency with `waitForUpdateUntilSlot` and `waitForUpdateUntilTime`

File: plutus-use-cases/src/Plutus/Contracts/Auction.hs [hunk 6]
```diff
@@ -243,10 +242,17 @@ auctionSeller value time = do
...
     :: StateMachineClient AuctionState AuctionInput
     -> Contract AuctionOutput BuyerSchema AuctionError (Maybe HighestBid)
-currentState client = mapError StateMachineContractError (SM.getOnChainState client) >>= \case
-    Just (SM.OnChainState{SM.ocsTxOut=TypedScriptTxOut{tyTxOutData=Ongoing s}}, _) -> do
-        tell $ auctionStateOut $ Ongoing s
-        pure (Just s)
+currentState client = do
+  mOcs <- mapError StateMachineContractError (SM.getOnChainState client)
+  case mOcs of
+    Just (ocs, _) -> do
```
Comment (sjoerdvisscher) [inline plutus-use-cases/src/Plutus/Contracts/Auction.hs:248]: Maybe use ViewPatterns here so you don't have to duplicate the failure case?

File: plutus-script-utils/src/Plutus/Script/Utils/V1/Typed/Scripts.hs [hunk 7]
```diff
@@ -26,3 +57,205 @@ The easiest way to tell is by reducing it to the previous problem: apply a funct
...
+      throwError $ WrongOutType ExpectedScriptGotPubkey
+    ScriptCredential _vh ->
+      case txOutDatumHash txOut of
+        Just dh | datumHash datum == dh -> do
+          checkValidatorAddress tv (txOutAddress txOut)
+          dsVal <- checkDatum tv datum
+          pure $ TypedScriptTxOut @out txOut dsVal
+        _ -> throwError $ NoDatum txOutRef (datumHash datum)
+
+-- | Create a 'TypedScriptTxOut' from an existing 'TxOut' by checking the types of its parts. To do this we
```
Comment (sjoerdvisscher) [inline plutus-script-utils/src/Plutus/Script/Utils/V1/Typed/Scripts.hs:245]: The comment is outdated.

**Review by sjoerdvisscher — APPROVED** [review]

File: plutus-use-cases/src/Plutus/Contracts/Auction.hs (hunk 6 above)
Comment (andreabedini) [reply plutus-use-cases/src/Plutus/Contracts/Auction.hs:248]: Good idea.

File: plutus-contract/src/Plutus/Contract/StateMachine.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (koslambrou) [reply plutus-contract/src/Plutus/Contract/StateMachine.hs:283]: :+1:
